storage_controller/src/http.rs

```diff
@@ -89,6 +91,46 @@ fn get_state(request: &Request<Body>) -> &HttpState {
         .as_ref()
 }
 
+/// Rate limits tenant requests.
+///
+/// TODO: this should be a request middleware, but requires us to extract the tenant ID from
+/// different URLs in a systematic way.
+///
+/// TODO: consider returning a 429 response if these start piling up.
+///
+/// TODO: consider adding a config option for the rate.
+async fn maybe_rate_limit(tenant_id: TenantId) {
+    // Check if the tenant should be rate-limited.
+    const TENANT_RATE_LIMIT: governor::Quota =
+        governor::Quota::per_second(NonZero::new(10).unwrap());
+    static TENANT_RATE_LIMITER: LazyLock<governor::DefaultKeyedRateLimiter<TenantId>> =
+        LazyLock::new(|| governor::RateLimiter::keyed(TENANT_RATE_LIMIT));
+
+    if TENANT_RATE_LIMITER.check_key(&tenant_id).is_ok() {
+        return;
+    }
+
+    // Log rate limited tenants once every 10 seconds.
+    static LOG_RATE_LIMITER: LazyLock<governor::DefaultKeyedRateLimiter<TenantId>> =
+        LazyLock::new(|| {
+            let quota = governor::Quota::with_period(Duration::from_secs(10)).unwrap();
+            governor::RateLimiter::keyed(quota)
+        });
+
+    if LOG_RATE_LIMITER.check_key(&tenant_id).is_ok() {
+        warn!("tenant {tenant_id} is rate limited")
+    }
+
+    // Measure the rate limiting delay.
+    let _timer = METRICS_REGISTRY
+        .metrics_group
+        .storage_controller_http_request_rate_limited
+        .start_timer();
+
+    // Wait for quota.
+    TENANT_RATE_LIMITER.until_key_ready(&tenant_id).await;
```

Comment:
It looks to me like the request futures are dropped on shutdown, so we don't need any explicit cancellation handling here.